packages/components/tree-v2/src/tree.vue

```diff
@@ -38,99 +38,87 @@
   </div>
 </template>
 
-<script lang="ts">
-import { defineComponent, getCurrentInstance, provide } from 'vue'
+<script lang="ts" setup>
+import { getCurrentInstance, provide, unref, useSlots } from 'vue'
 import { useLocale, useNamespace } from '@element-plus/hooks'
 import { formItemContextKey } from '@element-plus/tokens'
 import { FixedSizeList } from '@element-plus/components/virtual-list'
+import { useAttrs } from '@element-plus/hooks/use-attrs/index'
 import { useTree } from './composables/useTree'
 import ElTreeNode from './tree-node.vue'
 import { ROOT_TREE_INJECTION_KEY, treeEmits, treeProps } from './virtual-tree'
-import type { TreeProps } from './types'
 
-export default defineComponent({
+defineOptions({
   name: 'ElTreeV2',
-  components: {
-    ElTreeNode,
-    FixedSizeList,
-  },
-  props: treeProps,
-  emits: treeEmits,
-  setup(props: TreeProps, ctx) {
-    provide(ROOT_TREE_INJECTION_KEY, {
-      ctx,
-      props,
-      instance: getCurrentInstance(),
-    })
-    provide(formItemContextKey, undefined)
-    const { t } = useLocale()
-    const ns = useNamespace('tree')
-    const {
-      flattenTree,
-      isNotEmpty,
-      toggleExpand,
-      isExpanded,
-      isIndeterminate,
-      isChecked,
-      isDisabled,
-      isCurrent,
-      isForceHiddenExpandIcon,
-      toggleCheckbox,
-      handleNodeClick,
-      handleNodeCheck,
-      // expose
-      getCurrentNode,
-      getCurrentKey,
-      setCurrentKey,
-      getCheckedKeys,
-      getCheckedNodes,
-      getHalfCheckedKeys,
-      getHalfCheckedNodes,
-      setChecked,
-      setCheckedKeys,
-      filter,
-      setData,
-      getNode,
-      expandNode,
-      collapseNode,
-      setExpandedKeys,
-    } = useTree(props, ctx.emit)
+})
+
+const props = defineProps(treeProps)
+const emit = defineEmits(treeEmits)
 
-    ctx.expose({
-      getCurrentNode,
-      getCurrentKey,
-      setCurrentKey,
-      getCheckedKeys,
-      getCheckedNodes,
-      getHalfCheckedKeys,
-      getHalfCheckedNodes,
-      setChecked,
-      setCheckedKeys,
-      filter,
-      setData,
-      getNode,
-      expandNode,
-      collapseNode,
-      setExpandedKeys,
-    })
+const slots = useSlots()
+const attrs = useAttrs()
 
-    return {
-      t,
-      ns,
-      flattenTree,
-      itemSize: 26,
-      isNotEmpty,
-      toggleExpand,
-      toggleCheckbox,
-      isExpanded,
-      isIndeterminate,
-      isChecked,
-      isDisabled,
-      isCurrent,
-      isForceHiddenExpandIcon,
-      handleNodeClick,
-      handleNodeCheck,
-    }
+const itemSize = 26
+
+provide(ROOT_TREE_INJECTION_KEY, {
+  ctx: {
+    emit,
+    slots,
+    attrs: unref(attrs),
```

Comment:
Maybe don't need to pass `attrs`.

```suggestion
  ctx: {
    emit,
    slots,
```